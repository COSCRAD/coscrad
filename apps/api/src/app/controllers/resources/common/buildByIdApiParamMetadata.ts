export default () => ({
    name: 'id',
    required: true,
    example: '2',
});
