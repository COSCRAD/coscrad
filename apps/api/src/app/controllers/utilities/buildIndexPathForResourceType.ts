import { ResourceType } from '../../../domain/types/ResourceType';
import { RESOURCES_ROUTE_PREFIX } from '../resources/constants';

// For irregular plural forms or forms that lead to awkward grammar
const exceptions: Partial<Record<ResourceType, string>> = {
    [ResourceType.transcribedAudio]: 'transcribedAudioItems',
};

// `${baseResourcesPath}/${buildViewModelPathForResourceType(resourceType)}`

const pluralizeResourceTypeInCamelCase = (resourceType: ResourceType): string =>
    exceptions[resourceType] || `${resourceType}s`;

export default (resourceType: ResourceType): string =>
    `${RESOURCES_ROUTE_PREFIX}/${pluralizeResourceTypeInCamelCase(resourceType)}`;
