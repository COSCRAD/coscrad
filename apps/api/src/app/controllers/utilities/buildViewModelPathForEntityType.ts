import { EntityType } from 'apps/api/src/domain/types/entityType';

// We can deal with irregular plurals when that becomes a problem.
export default (entityType: EntityType): string => `${entityType}s`;
