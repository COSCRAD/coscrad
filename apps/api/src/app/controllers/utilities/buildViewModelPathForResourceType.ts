import { ResourceType, resourceTypes } from '../../../domain/types/resourceTypes';

// For irregular plural forms or forms that lead to awkward grammar
const exceptions: Partial<Record<ResourceType, string>> = {
    [resourceTypes.transcribedAudio]: 'transcribedAudioItems',
};

export default (resourceType: ResourceType): string =>
    exceptions[resourceType] || `${resourceType}s`;
