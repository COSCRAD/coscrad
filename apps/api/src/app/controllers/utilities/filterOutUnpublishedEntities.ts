export type HasPublicationStatus = {
  isPublished: boolean;
};

export default <T extends HasPublicationStatus>(input: T[]): T[] =>
  input.filter((publishableItem) => publishableItem.isPublished);
