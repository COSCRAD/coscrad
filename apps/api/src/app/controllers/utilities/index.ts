// TODO Barrel export everything in this directory and fix existing imports
export * from './mixLinkIntoViewModelDescription';
