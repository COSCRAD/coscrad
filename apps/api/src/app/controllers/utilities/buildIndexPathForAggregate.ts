import { isStringWithNonzeroLength } from '@coscrad/validation';
import { AggregateType } from '../../../domain/types/AggregateType';
import { isResourceType, ResourceType } from '../../../domain/types/ResourceType';
import { InternalError } from '../../../lib/errors/InternalError';
import {
    CATEGORY_TREE_INDEX_ROUTE,
    NOTE_INDEX_ROUTE,
    USER_GROUP_INDEX_ROUTE,
    USER_INDEX_ROUTE,
} from '../constants';
import { TAG_INDEX_ROUTE } from '../tag.controller';
import buildIndexPathForResourceType from './buildIndexPathForResourceType';

const nonResourceAggregateTypeToIndexPath: {
    [K in Exclude<AggregateType, ResourceType>]: string;
} = {
    note: NOTE_INDEX_ROUTE,
    tag: TAG_INDEX_ROUTE,
    category: CATEGORY_TREE_INDEX_ROUTE,
    user: USER_INDEX_ROUTE,
    userGroup: USER_GROUP_INDEX_ROUTE,
};

export const buildIndexPathForAggregate = (aggregateType: AggregateType): string => {
    if (isResourceType(aggregateType)) return buildIndexPathForResourceType(aggregateType);

    const lookupResult = nonResourceAggregateTypeToIndexPath[aggregateType];

    if (!isStringWithNonzeroLength(lookupResult)) {
        throw new InternalError(
            `There is no route registered for aggregate of type: ${aggregateType}`
        );
    }

    return lookupResult;
};
