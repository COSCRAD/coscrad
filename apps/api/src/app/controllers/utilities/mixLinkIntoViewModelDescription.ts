import { AggregateInfo } from '../../../view-models/resourceDescriptions/types/AggregateInfo';
import { buildIndexPathForAggregate } from './buildIndexPathForAggregate';

/**
 * Note that this is curried because we almost always want
 * to generate several such links for one single global prefix
 */
export const mixLinkIntoViewModelDescription =
    /**
     * We don't actually need to mix in the global prefix. This is part of the
     * config on the client anyway. We should refactor to no longer curry this.
     */


        (_: string) =>
        (aggregateInfo: Omit<AggregateInfo, 'link'>): AggregateInfo => ({
            ...aggregateInfo,
            link: `/${buildIndexPathForAggregate(aggregateInfo.type)}`,
        });
