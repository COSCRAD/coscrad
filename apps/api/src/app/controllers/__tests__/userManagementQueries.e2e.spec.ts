import { CoscradUserRole } from '@coscrad/data-types';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { CoscradUserWithGroups } from '../../../domain/models/user-management/user/entities/user/coscrad-user-with-groups';
import { dummySystemUserId } from '../../../domain/models/__tests__/utilities/dummySystemUserId';
import buildInMemorySnapshot from '../../../domain/utilities/buildInMemorySnapshot';
import generateDatabaseNameForTestSuite from '../../../persistence/repositories/__tests__/generateDatabaseNameForTestSuite';
import TestRepositoryProvider from '../../../persistence/repositories/__tests__/TestRepositoryProvider';
import buildTestData from '../../../test-data/buildTestData';
import httpStatusCodes from '../../constants/httpStatusCodes';
import setUpIntegrationTest from './setUpIntegrationTest';
import assertDetailQueryOkResponse from './sharedAssertions/assertDetailQueryOkResponse';
import assertIndexQueryOkResponse from './sharedAssertions/assertIndexQueryOkResponse';

describe('User Management Queries', () => {
    const testDatabaseName = generateDatabaseNameForTestSuite();

    const userGroupIndexEndpoint = '/userGroups';

    const userIndexEndpoint = '/users';

    const { userGroup: userGroups, user: users } = buildTestData();

    const popularUser = users[0];

    const projectAdminUser = popularUser.clone({
        id: dummySystemUserId,
        username: 'project-admin',
        authProviderUserId: 'auth0|86868686868',
        roles: [CoscradUserRole.projectAdmin],
    });

    const coscradAdminUser = popularUser.clone({
        id: dummySystemUserId,
        username: 'coscrad-system-admin',
        authProviderUserId: 'auth0|86868686855',
        roles: [CoscradUserRole.superAdmin],
    });

    const nonAdminSystemUser = popularUser.clone({
        id: dummySystemUserId,
        username: 'ordinary-user',
        authProviderUserId: 'auth0|86868686859',
        roles: [CoscradUserRole.viewer],
    });

    const snapshotWithUsersAndGroups = buildInMemorySnapshot({
        userGroup: userGroups.map((group) =>
            group.clone({
                userIds: [popularUser.id],
            })
        ),
        user: users,
    });

    describe(`when fetching schemas (/admin/)`, () => {
        let app: INestApplication;

        let testRepositoryProvider: TestRepositoryProvider;
        beforeAll(async () => {
            ({ app, testRepositoryProvider } = await setUpIntegrationTest({
                ARANGO_DB_NAME: testDatabaseName,
            }));
        });

        afterAll(async () => {
            await app.close();
        });

        beforeEach(async () => {
            await testRepositoryProvider.testSetup();
        });
        it('should return the expected result', async () => {
            const result = await request(app.getHttpServer()).get('/admin');

            expect(result.status).toBe(httpStatusCodes.ok);

            expect(result.body).toMatchSnapshot();
        });
    });

    describe('when the system user issuing the command is authorized', () => {
        (
            [
                [
                    new CoscradUserWithGroups(coscradAdminUser, []),
                    'when the user is a coscrad admin',
                ],
                [
                    new CoscradUserWithGroups(projectAdminUser, []),
                    'when the user is a project admin',
                ],
            ] as const
        ).forEach(([adminUserWithGroups, description]) => {
            describe(description, () => {
                const testDatabaseName = generateDatabaseNameForTestSuite();

                let app: INestApplication;

                let testRepositoryProvider: TestRepositoryProvider;
                beforeAll(async () => {
                    ({ app, testRepositoryProvider } = await setUpIntegrationTest(
                        {
                            ARANGO_DB_NAME: testDatabaseName,
                        },
                        {
                            testUserWithGroups: adminUserWithGroups,
                        }
                    ));
                });

                afterAll(async () => {
                    await app.close();
                });

                beforeEach(async () => {
                    await testRepositoryProvider.testSetup();
                });

                afterEach(async () => {
                    await testRepositoryProvider.testTeardown();
                });

                describe('when querying for user groups', () => {
                    describe('when fetching a group by ID', () => {
                        describe('when there is no group with the given ID', () => {
                            it('should return not found', async () => {
                                await testRepositoryProvider.addFullSnapshot(
                                    snapshotWithUsersAndGroups
                                );

                                return await request(app.getHttpServer())
                                    .get(`${userGroupIndexEndpoint}/'BOGUS-GROUP-ID`)
                                    .expect(httpStatusCodes.notFound);
                            });
                        });

                        describe('when there is a group with the given ID', () => {
                            it('should find the group', async () => {
                                await testRepositoryProvider.addFullSnapshot(
                                    snapshotWithUsersAndGroups
                                );

                                const idToFind = userGroups[0].id;

                                const result = await request(app.getHttpServer()).get(
                                    `${userGroupIndexEndpoint}/${idToFind}`
                                );

                                assertDetailQueryOkResponse(result, idToFind);
                            });
                        });
                    });

                    describe('when fetching all groups', () => {
                        it('should return the expected result', async () => {
                            await testRepositoryProvider.addFullSnapshot(
                                snapshotWithUsersAndGroups
                            );

                            const res = await request(app.getHttpServer()).get(
                                userGroupIndexEndpoint
                            );

                            assertIndexQueryOkResponse(res);
                        });
                    });
                });

                describe('when querying for users', () => {
                    describe('when fetching a user by ID', () => {
                        describe('when there is no user with the given ID', () => {
                            it('should return not found', async () => {
                                await testRepositoryProvider.addFullSnapshot(
                                    snapshotWithUsersAndGroups
                                );

                                return await request(app.getHttpServer())
                                    .get(`${userIndexEndpoint}/'BOGUS-USER-ID`)
                                    .expect(httpStatusCodes.notFound);
                            });
                        });

                        describe('when there is a user with the given ID', () => {
                            it('should find the user', async () => {
                                await testRepositoryProvider.addFullSnapshot(
                                    snapshotWithUsersAndGroups
                                );

                                const idToFind = users[0].id;

                                const result = await request(app.getHttpServer()).get(
                                    `${userIndexEndpoint}/${idToFind}`
                                );

                                assertDetailQueryOkResponse(result, idToFind);
                            });
                        });
                    });

                    describe('when fetching all users', () => {
                        it('should return the expected result', async () => {
                            await testRepositoryProvider.addFullSnapshot(
                                snapshotWithUsersAndGroups
                            );

                            const res = await request(app.getHttpServer()).get(userIndexEndpoint);

                            assertIndexQueryOkResponse(res);
                        });
                    });
                });
            });
        });
    });

    describe('when the system user issuing the command is not authorized', () => {
        (
            [
                [
                    new CoscradUserWithGroups(nonAdminSystemUser, userGroups),
                    'when the user is authenticated as an ordinary, non-admin user',
                ],
                [undefined, 'when the user is not logged in'],
            ] as const
        ).forEach(([unauthorizedSystemUserWithGroups, description]) => {
            describe(description, () => {
                let app: INestApplication;

                let testRepositoryProvider: TestRepositoryProvider;
                beforeAll(async () => {
                    ({ app, testRepositoryProvider } = await setUpIntegrationTest(
                        {
                            ARANGO_DB_NAME: testDatabaseName,
                        },
                        {
                            testUserWithGroups: unauthorizedSystemUserWithGroups,
                        }
                    ));
                });

                afterAll(async () => {
                    await app.close();
                });

                beforeEach(async () => {
                    await testRepositoryProvider.testSetup();
                });

                afterEach(async () => {
                    await testRepositoryProvider.testTeardown();
                });

                describe('when querying for a single group by ID (fetchById)', () => {
                    it('should return forbidden', async () => {
                        await testRepositoryProvider.addFullSnapshot(snapshotWithUsersAndGroups);

                        const idToFind = userGroups[0].id;

                        return await request(app.getHttpServer())
                            .get(`${userGroupIndexEndpoint}/${idToFind}`)
                            .expect(httpStatusCodes.forbidden);
                    });
                });

                describe('when querying for many groups (fetchMany)', () => {
                    it('should return forbidden', async () => {
                        await testRepositoryProvider.addFullSnapshot(snapshotWithUsersAndGroups);

                        await request(app.getHttpServer())
                            .get(userGroupIndexEndpoint)
                            .expect(httpStatusCodes.forbidden);
                    });
                });

                describe('when querying for a single user by ID (fetchById)', () => {
                    it('should return forbidden', async () => {
                        await testRepositoryProvider.addFullSnapshot(snapshotWithUsersAndGroups);

                        const idToFind = users[0].id;

                        return await request(app.getHttpServer())
                            .get(`${userIndexEndpoint}/${idToFind}`)
                            .expect(httpStatusCodes.forbidden);
                    });
                });

                describe('when querying for many users (fetchMany)', () => {
                    it('should return forbidden', async () => {
                        await testRepositoryProvider.addFullSnapshot(snapshotWithUsersAndGroups);

                        await request(app.getHttpServer())
                            .get(userIndexEndpoint)
                            .expect(httpStatusCodes.forbidden);
                    });
                });
            });
        });
    });
});
