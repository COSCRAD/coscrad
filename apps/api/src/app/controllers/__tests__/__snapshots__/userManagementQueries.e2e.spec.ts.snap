// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Group Queries when the system user issuing the command is authorized when the user is a coscrad admin when querying for user groups when fetching a group by ID when there is a group with the given ID should find the group 1`] = `
Object {
  "actions": Array [],
  "data": Object {
    "description": "This group is for members of the community only.",
    "id": "1",
    "label": "community members",
    "users": Array [
      Object {
        "id": "1",
        "profile": Object {
          "email": "llcj1985@aol.com",
          "name": Object {
            "firstName": "James",
            "lastName": "Jones",
          },
        },
        "roles": Array [
          "viewer",
        ],
        "username": "cool-james",
      },
    ],
  },
}
`;

exports[`User Group Queries when the system user issuing the command is authorized when the user is a coscrad admin when querying for user groups when fetching all groups should return the expected result 1`] = `
Object {
  "actions": Array [],
  "data": Array [
    Object {
      "actions": Array [],
      "data": Object {
        "description": "This group is for members of the community only.",
        "id": "1",
        "label": "community members",
        "users": Array [
          Object {
            "id": "1",
            "profile": Object {
              "email": "llcj1985@aol.com",
              "name": Object {
                "firstName": "James",
                "lastName": "Jones",
              },
            },
            "roles": Array [
              "viewer",
            ],
            "username": "cool-james",
          },
        ],
      },
    },
  ],
}
`;

exports[`User Group Queries when the system user issuing the command is authorized when the user is a coscrad admin when querying for users when fetching a user by ID when there is a user with the given ID should find the user 1`] = `
Object {
  "actions": Array [],
  "data": Object {
    "id": "1",
    "profile": Object {
      "email": "llcj1985@aol.com",
      "name": Object {
        "firstName": "James",
        "lastName": "Jones",
      },
    },
    "roles": Array [
      "viewer",
    ],
    "username": "cool-james",
  },
}
`;

exports[`User Group Queries when the system user issuing the command is authorized when the user is a coscrad admin when querying for users when fetching all users should return the expected result 1`] = `
Object {
  "actions": Array [],
  "data": Array [
    Object {
      "actions": Array [],
      "data": Object {
        "id": "1",
        "profile": Object {
          "email": "llcj1985@aol.com",
          "name": Object {
            "firstName": "James",
            "lastName": "Jones",
          },
        },
        "roles": Array [
          "viewer",
        ],
        "username": "cool-james",
      },
    },
  ],
}
`;

exports[`User Group Queries when the system user issuing the command is authorized when the user is a project admin when querying for user groups when fetching a group by ID when there is a group with the given ID should find the group 1`] = `
Object {
  "actions": Array [],
  "data": Object {
    "description": "This group is for members of the community only.",
    "id": "1",
    "label": "community members",
    "users": Array [
      Object {
        "id": "1",
        "profile": Object {
          "email": "llcj1985@aol.com",
          "name": Object {
            "firstName": "James",
            "lastName": "Jones",
          },
        },
        "roles": Array [
          "viewer",
        ],
        "username": "cool-james",
      },
    ],
  },
}
`;

exports[`User Group Queries when the system user issuing the command is authorized when the user is a project admin when querying for user groups when fetching all groups should return the expected result 1`] = `
Object {
  "actions": Array [],
  "data": Array [
    Object {
      "actions": Array [],
      "data": Object {
        "description": "This group is for members of the community only.",
        "id": "1",
        "label": "community members",
        "users": Array [
          Object {
            "id": "1",
            "profile": Object {
              "email": "llcj1985@aol.com",
              "name": Object {
                "firstName": "James",
                "lastName": "Jones",
              },
            },
            "roles": Array [
              "viewer",
            ],
            "username": "cool-james",
          },
        ],
      },
    },
  ],
}
`;

exports[`User Group Queries when the system user issuing the command is authorized when the user is a project admin when querying for users when fetching a user by ID when there is a user with the given ID should find the user 1`] = `
Object {
  "actions": Array [],
  "data": Object {
    "id": "1",
    "profile": Object {
      "email": "llcj1985@aol.com",
      "name": Object {
        "firstName": "James",
        "lastName": "Jones",
      },
    },
    "roles": Array [
      "viewer",
    ],
    "username": "cool-james",
  },
}
`;

exports[`User Group Queries when the system user issuing the command is authorized when the user is a project admin when querying for users when fetching all users should return the expected result 1`] = `
Object {
  "actions": Array [],
  "data": Array [
    Object {
      "actions": Array [],
      "data": Object {
        "id": "1",
        "profile": Object {
          "email": "llcj1985@aol.com",
          "name": Object {
            "firstName": "James",
            "lastName": "Jones",
          },
        },
        "roles": Array [
          "viewer",
        ],
        "username": "cool-james",
      },
    },
  ],
}
`;
